pkg/gauges/contracts/gauges/PremintedGauge.sol

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ReentrancyGuard.sol";
+
+import "../interfaces/IBalancerMinter.sol";
+import "../interfaces/IBalancerTokenAdmin.sol";
+import "../interfaces/IGaugeController.sol";
+import "../interfaces/ILiquidityGauge.sol";
+
+abstract contract PremintedGauge is ILiquidityGauge, ReentrancyGuard {
+    IERC20 internal immutable _balToken;
+    IBalancerTokenAdmin private immutable _tokenAdmin;
+    IBalancerMinter private immutable _minter;
+    IGaugeController private immutable _gaugeController;
+
+    uint256 private immutable _INITIAL_RATE;
+    uint256 private immutable _RATE_REDUCTION_TIME;
+    uint256 private immutable _RATE_REDUCTION_COEFFICIENT;
+    uint256 private immutable _RATE_DENOMINATOR;
+
+    uint256 private _rate;
+    uint256 private _period;
+    uint256 private _startEpochTime;
+
+    uint256 private _emissions;
+    bool public isKilled;
+
+    constructor(IBalancerMinter minter) {
+        IBalancerTokenAdmin tokenAdmin = IBalancerTokenAdmin(minter.getBalancerTokenAdmin());
+        IERC20 balToken = tokenAdmin.getBalancerToken();
+
+        _balToken = balToken;
+        _tokenAdmin = tokenAdmin;
+        _minter = minter;
+        _gaugeController = minter.getGaugeController();
+
+        _INITIAL_RATE = tokenAdmin.INITIAL_RATE();
+        _RATE_REDUCTION_TIME = tokenAdmin.RATE_REDUCTION_TIME();
+        _RATE_REDUCTION_COEFFICIENT = tokenAdmin.RATE_REDUCTION_COEFFICIENT();
+        _RATE_DENOMINATOR = tokenAdmin.RATE_DENOMINATOR();
+
+        // Because we calculate the rate locally, this gauge cannot
+        // be used prior to the start of the first emission period
+        uint256 rate = tokenAdmin.rate();
+        require(rate != 0);
```

Comment:
This would be `0xffffff...fff` in our case, no?

Reply:
We have `rate = 0` when the `tokenAdmin` is unactivated so this is correct. I think you're mixing this up with `startEpochTime` being `0xffffff...fff`

Reply:
You're right, thanks!